pkg/ext/stores/tokens/tokens.go

```diff
@@ -0,0 +1,844 @@
+	hasPermission, err := t.support.UserHasManageTokenPermissions(ctx)
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to check user permissions: %w", err))
+	}
+
+	ch := make(chan extcore.WatchEvent[*ext.Token])
+
+	go func() {
+		// watch backend secret, transform its events into token events
+		watcher, err := t.secretClient.Watch(TokenNamespace, metav1.ListOptions{
+			ResourceVersion: opts.ResourceVersion,
+		})
+		if err != nil {
+			close(ch)
+			return
+		}
+
+		ctxUser := ctx.User.GetName()
+
+		defer watcher.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				// context got cancelled - done
+				close(ch)
+				return
+			case event, more := <-watcher.ResultChan():
+				// no more events - done
+				if !more {
+					close(ch)
+					return
+				}
+
+				secret, ok := event.Object.(*corev1.Secret)
+				if !ok {
+					continue
+				}
+
+				token, err := tokenFromSecret(secret)
+				if err != nil {
+					// ignore broken tokens
+					continue
+				}
+				if !hasPermission && token.Spec.UserID != ctxUser {
+					// ignore tokens belonging to other users if this user does
+					// not have the necessary permissions
+					continue
+				}
+
+				ch <- extcore.WatchEvent[*ext.Token]{
+					Event:  event.Type,
+					Object: token,
+				}
+			}
+		}
+	}()
+
+	return ch, nil
+}
+
+func (t *TokenStore) Delete(ctx extcore.Context, name string, opts *metav1.DeleteOptions) error {
+	// have to pull the token information first before we can check permissions
+	secret, err := t.secretClient.Get(TokenNamespace, name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return err
+		}
+		return apierrors.NewInternalError(fmt.Errorf("failed to retrieve token %s: %w", name, err))
+	}
+
+	// (**) conversion errors are ignored here. we still get a partially filled token usable
+	// enough for the permission check. even a missing user id (== empty string) is
+	// acceptable. this simply triggers the non-owned secret code path, preventing deletion from
+	// all but token admins.
+	token, _ := tokenFromSecret(secret)
+	if _, err := t.support.UserHasPermission("delete", token, ctx); err != nil {
+		return err
+	}
+
+	return t.SystemTokenStore.Delete(name, opts)
+}
+
+func (t *SystemTokenStore) Delete(name string, opts *metav1.DeleteOptions) error {
+	err := t.secretClient.Delete(TokenNamespace, name, opts)
+	if err == nil {
+		return nil
+	}
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	return apierrors.NewInternalError(fmt.Errorf("failed to delete token %s: %w", name, err))
+}
+
+// Mockable setup for permission checking, secret generation and hashing, and timing
+
+// supportActionHandler is an interface hiding the details of permission checking
+// secret generation and hashing, and timeing from the store. This makes these
+// operations mockable for store testing.
+type supportActionHandler interface {
```

Comment:
IMHO we should avoid using types/objects with names that are too broad/generic like `util`,`support` etc. They inevitably become kitchen sinks of things that don't belong together (separation of concerns), which seems to be the case here dealing with permissions, secrets and time are all separate concerns.

Reply:
More over checking permissions is a common concern for ALL ext endpoints and ideally should be reusable.

Reply:
Yeah, this was and is a quick hack to make this functionality mockable in the tests.
Note that splitting this creates 2 additional arguments in the low-level constructors (`New(System)`).
It luckily does not escape into the rest of the system. (Hidden in `New...FromWrangler`)

Can we defer this until we have multiple ext stores ?

Because I can then see this move outside of this package to be sharable.

Reply:
Done.